pkg/webhook/resourcedistribution/validating/resourcedistribution_create_update_handler.go

```diff
@@ -0,0 +1,184 @@
+	if resource == nil {
+		return
+	}
+
+	// 2. deserialize old resource if need
+	var oldResource runtime.Object = nil
+	if oldObj != nil {
+		oldResource, errs = DeserializeResource(&oldObj.Spec.Resource, fldPath)
+		allErrs = append(allErrs, errs...)
+	}
+
+	// 3. validate resource
+	allErrs = append(allErrs, h.validateResourceDistributionResource(resource, oldResource, fldPath.Child("resource"))...)
+
+	// 4. validate targets and conflict
+	allErrs = append(allErrs, h.validateResourceDistributionSpecTargets(obj, resource, fldPath.Child("targets"))...)
+	return
+}
+
+// validateResourceDistributionSpecResource validate Spec.Resource when creating and updating
+// (1). validate resource kind and semantics
+// (2). validate updating conflict, GVK and name cannot be modified
+func (h *ResourceDistributionCreateUpdateHandler) validateResourceDistributionResource(resource, oldResource runtime.Object, fldPath *field.Path) (allErrs field.ErrorList) {
+	// 1. validate resource kind and semantics
+	allErrs = append(allErrs, h.validateResourceDistributionResourceKindAndSemantics(resource, fldPath)...)
+
+	// 2. validate resource groupVersionKind and name when updating
+	if len(allErrs) == 0 && oldResource != nil {
+		newName, oldName := ConvertToUnstructured(resource).GetName(), ConvertToUnstructured(oldResource).GetName()
+		newGVK, oldGVK := resource.GetObjectKind().GroupVersionKind(), oldResource.GetObjectKind().GroupVersionKind()
+		if newName != oldName || !reflect.DeepEqual(newGVK, oldGVK) {
+			allErrs = append(allErrs, field.Invalid(fldPath, nil, "resource apiVersion, kind, and name are immutable"))
+		}
+	}
+	return
+}
+
+func (h *ResourceDistributionCreateUpdateHandler) validateResourceDistributionResourceKindAndSemantics(resource runtime.Object, fldPath *field.Path) (allErrs field.ErrorList) {
+	// 1. check whether the GVK of the resource is in supportGVKList
+	if !SupportResourceGVK(resource) {
+		return append(allErrs, field.Invalid(fldPath, resource, fmt.Sprintf("unknown or unsupported resource GroupVersionKind, only support %v", supportGVKList)))
+	}
+
+	// 2. dry run to check resource
+	mice := resource.DeepCopyObject()
+	ConvertToUnstructured(mice).SetNamespace(DefaultNamespace)
+	if err := h.Client.Create(context.TODO(), mice, &client.CreateOptions{
+		DryRun: []string{metav1.DryRunAll},
+	}); err != nil {
+		return append(allErrs, field.InternalError(fldPath, fmt.Errorf("create namespaces failed, err: %v", err)))
+	}
+	return
+}
+
+// validateResourceDistributionSpecTargets validate Spec.Targets
+// (1). parse target namespaces
+// (2). validate conflict between existing resources
+func (h *ResourceDistributionCreateUpdateHandler) validateResourceDistributionSpecTargets(obj *appsv1alpha1.ResourceDistribution, resource runtime.Object, fldPath *field.Path) (allErrs field.ErrorList) {
+	spec := &obj.Spec
+	// 1. parse target namespaces
+	targetNamespaces, errs := GetTargetNamespaces(h.Client, &spec.Targets)
```

Comment:
it is too costly to list namespaces and objects in a webhook. plz consider move the list operation into the controller, and do basic checking here, for example
1. ensure namespace name is valid
2. ensure nameppace label selector is valid

Reply:
ok